### PR TITLE
[Merl 377] Migration Guide

### PR DESCRIPTION
**Pull request by theodesp**
## Description

This PR adds the migration guide in the docs explaining the steps required to transition from old Faust to the new one.

## Testing

- Visit `docs/migrationPath/overview`
- Check That there is a link in the front page about the migration path.

## Screenshots
Screenshot 2022-11-15 at 11 45 58

### CODE REVIEW

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 1]
```diff
@@ -2,6 +2,268 @@
...
+  src="/docs/img/possible-templates.png"
+  alt="The list of possible templates as shown in the dev console"
+/>
+
+You can see an example of the hierarchy in the [blueprint project](https://github.com/wpengine/atlas-blueprint-portfolio/tree/main/wp-templates) where we provided implementations for the main templates.
+
+## Overview of the changes needed in to migrate from GQty to Apollo
+
+Given the above information, we can now list an brief overview of the changes needed in to migrate from GQty to Apollo.
+You can also review the code of this [Blueprint Project](https://github.com/wpengine/atlas-blueprint-portfolio/tree/main) that shows how the end result will look like.
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:129]:
```suggestion
:::tip
Take a look at the [Blueprint Portfolio](https://github.com/wpengine/atlas-blueprint-portfolio/tree/main) for an example of a migrated project in Faust.
:::
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 2]
```diff
@@ -2,6 +2,268 @@
...
+
+**1. Capture and migrate Presentational Components**
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:131]:
```suggestion
**1. Capture and Migrate Presentational Components**
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 3]
```diff
@@ -2,6 +2,268 @@
...
+
+Create a list of components that can be reused into the new version of Faust and move them into the new projects folder.
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:133]:
```suggestion
Create a list of components that can be reused in the new version of Faust and move them into the new projects folder.
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 4]
```diff
@@ -2,6 +2,268 @@
...
+
+**2. Import Faust Dependencies**
+
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:136]: Do we want folks to remove any old `node_modules` first before the import?

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 5]
```diff
@@ -2,6 +2,268 @@
...
+    "dev": "faust dev",
+    "build": "faust build",
+    "generate": "faust generatePossibleTypes",
+    "start": "faust start",
+    "lint": "faust lint"
+	...
+}
+```
+
+**3. Use new Faust Configuration**
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:154]:
```suggestion
**3. Use New Faust Configuration**
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 6]
```diff
@@ -2,6 +2,268 @@
...
+ * @type {import('@faustwp/core').FaustConfig}
+ **/
+export default setConfig({
+  templates,
+  experimentalPlugins: [],
+  possibleTypes,
+});
+```
+
+**4. Use new Faust Imports and Providers**
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:182]:
```suggestion
**4. Use new Faust Imports and Providers**

```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 7]
```diff
@@ -2,6 +2,268 @@
...
+    <>
+      <FaustProvider pageProps={pageProps}>
+        <Component {...pageProps} key={router.asPath} />
+      </FaustProvider>
+    </>
+  );
+}
+```
+
+**5. Create the new wp-templates folder and include an index**
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:214]:
```suggestion
**5. Create the new wp-templates folder and include an index**

```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 8]
```diff
@@ -2,6 +2,268 @@
...
+The blueprint example project showcases how the components are structured and it's a great way to understand what needs to be done.
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:215]:
```suggestion
:::tip
The blueprint example project showcases how the components are structured and is a great way to understand what changes to make during this step.
:::

```

Comment (TeresaGobble) [reply internal/faustjs.org/docs/migrationPath/overview.mdx:215]: A great suggestion! Is this Faust's blank scaffold Blake is working on this sprint? Or the portfolio?

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 9]
```diff
@@ -2,6 +2,268 @@
...
+You should be using the Apollo Client from now on to prepare queries and mutations.
+
+If you are not using the `wp-templates` system you can take a look a the [search.js](https://github.com/wpengine/atlas-blueprint-portfolio/blob/main/pages/search.js)
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:218]:
```suggestion
If you are not using the `wp-templates` system, you can take a look a the [search.js](https://github.com/wpengine/atlas-blueprint-portfolio/blob/main/pages/search.js)
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 10]
```diff
@@ -2,6 +2,268 @@
...
+
+Those include:
+
+* The `gqty.config.js`.
+* The `client` folder.
+* The `@faustjs/core`, `@faustjs/next` and `gqty` dependencies.
+* Any [skeleton renders](https://gqty.dev/docs/intro/how-it-works#interplay-with-lists--keys) as a result of GQty side effects.
+
+
+## Any new tools helpful in working with Apollo
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:261]:
```suggestion
## Helpful Tools When Working with Apollo
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 11]
```diff
@@ -2,6 +2,268 @@
...
+
+While migrating your app from GQty to Apollo, you will find Apollo Ecosystem has plenty of resources and tools to help you work with GraphQL.
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:263]:
```suggestion
While migrating your app from GQty to Apollo, you will find that the Apollo Ecosystem has plenty of resources and tools to help you work with GraphQL.
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 12]
```diff
@@ -2,6 +2,268 @@
...
+We list some of the available tools here that will definitely help you:
+
+* **Apollo Client DevTools**: DevTools is a debugger toolbar for Chrome that allows you to inspect and perform queries against the GraphQL Endpoint within the browser.
+    This tools is invaluable when trying to migrate existing GQty hooks into Apollo.
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:267]:
```suggestion
    This tool is invaluable when trying to migrate existing GQty hooks into Apollo.
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 13]
```diff
@@ -2,6 +2,268 @@
...
+    [Link to Download](https://chrome.google.com/webstore/detail/apollo-client-devtools/jdkknkkbebbapilgoeccciglkfbmbnfm)
+* **Apollo Fragments Guide**: Using the new Faust, you will have to write queries manually from now on. Using GraphQL fragments is a recommended way to share pieces of logic between multiple queries and mutations. Read [this guide](https://www.apollographql.com/docs/react/data/fragments/) for best practices and greate techniques to maximize using fragments.
```
Comment (TeresaGobble) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:269]:
```suggestion
* **Apollo Fragments Guide**: Using the new Faust, you will need to write queries manually. Using GraphQL fragments is a recommended way to share pieces of logic between multiple queries and mutations. Read [this guide](https://www.apollographql.com/docs/react/data/fragments/) for best practices and great techniques to maximize using fragments.
```

**Review by TeresaGobble — CHANGES_REQUESTED** [review]
Awesome job, Theo! This will be super helpful to folks as they begin their migration process!

File: internal/faustjs.org/docs/migrationPath/overview.mdx (hunk 8 above)
Comment (TeresaGobble) [reply internal/faustjs.org/docs/migrationPath/overview.mdx:216]: If we can link to whichever blueprint this one references, I think that will be really helpful for clarity

**Review by TeresaGobble — CHANGES_REQUESTED** [review]
Hi Theo! I have just one note on adding the link to the referenced blueprint to clarify which is referred to here. Great work!

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 14]
```diff
@@ -2,6 +2,269 @@
...
+Currently, there is no great automated way to migrate from previous versions of Faust that use GQty to the recent version using Apollo. This is primarily due to their striking differences in the way the queries are performed.
+However, there are some conventional techniques and best practices for React Development that will definitely help you with this process.
+
+### Reusing Presentational Components
+If you paid attention while developing your Faust website, the Components folder typically contains components that embody a presentational meaning.
+
+By presentational we mean they do not depend on GQty or any hooks or side-effects and they just take props and render the data.
+Here is an example of a presentational component taken from this [blueprint](https://github.com/wpengine/atlas-blueprint-portfolio/blob/main/components/PostInfo/PostInfo.js):
+
+```js title="PostInfo.js"
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:17]:
```suggestion
```jsx title="PostInfo.js"
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 15]
```diff
@@ -2,6 +2,269 @@
...
+        </time>
+      )}
+      {date && author && <>&nbsp;</>}
+      {author && <span>By {author}</span>}
+    </div>
+  );
+}
+```
+
+There is no reference to GQty or Apollo and this component can be safely re-used accross different applications.
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:37]:
```suggestion
There is no reference to GQty or Apollo and this component can be safely re-used across different applications.
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 16]
```diff
@@ -2,6 +2,269 @@
...
+    title
+    author {
+    	node {
+        	name
+    	}
+    }
+}
+```
+
+This fragment is reusable and can be applied to any page query when requsting the Post data fields:
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:87]:
```suggestion
This fragment is reusable and can be applied to any page query when requesting the Post data fields:
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 17]
```diff
@@ -2,6 +2,269 @@
...
+
+```js title="pages/preview.js"
+import { getWordPressProps, WordPressTemplate } from '@faustwp/core';
+
+export default function Page(props) {
+  return <WordPressTemplate {...props} />;
+}
+```
+
+At this point you should be able to run the dev server and inpsect if something is missing or not working properly.
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:248]:
```suggestion
At this point you should be able to run the dev server and inspect if something is missing or not working properly.
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 18]
```diff
@@ -2,6 +2,269 @@
...
+    </div>
+  );
+}
+```
+
+There is no reference to GQty or Apollo and this component can be safely re-used accross different applications.
+
+On the other hand the following component is harder to migrate as it depends on the GQty client:
+
+```js title="NavigationMenu.js"
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:41]:
```suggestion
```jsx title="NavigationMenu.js"
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 19]
```diff
@@ -2,6 +2,269 @@
...
+  const { nodes: menuItems } = useQuery().menuItems({
+    where: {
+      location: menuLocation,
+    },
+  });
+  ...
+```
+You will have to either replace the `useQuery()` from GQty with the `useQuery()` of Apollo or perform the query on a higher level and pass on the menu items as props:
+
+```js title="NavigationMenu.js"
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:53]:
```suggestion
```jsx title="NavigationMenu.js"
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 20]
```diff
@@ -2,6 +2,269 @@
...
+If you developed your components using this approach, you will be able to just transfer them without modifications to the new framework.
+
+### Writing GraphQL Queries
+The biggest hurdle when migrating from GQty to Apollo is that now you will need to write queries manually.
+However this is not as bad as you think. You only need to identify the data that your pages and components need and try to request the appropriate fields.
+
+For instance take a look at the `PostInfo` component we posted before. It is supposed to display the `Post` Information using the `author` and `date` fields.
+You can request those fields once in the post or page fragment:
+
+```
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:73]:
```suggestion
```graphql
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 21]
```diff
@@ -2,6 +2,269 @@
...
+    	node {
+        	name
+    	}
+    }
+}
+```
+
+This fragment is reusable and can be applied to any page query when requsting the Post data fields:
+
+```
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:89]:
```suggestion
```graphql
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 22]
```diff
@@ -2,6 +2,269 @@
...
+For the api route:
+
+```js title="pages/api/faust/[[...route]].js"
+// import { apiRouter } from '@faustjs/core/api';
+import { apiRouter } from '@faustwp/core';
+```
+
+For the `_app.js`:
+
+```js title="_app.js"
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:195]:
```suggestion
```jsx title="_app.js"
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 23]
```diff
@@ -2,6 +2,269 @@
...
+You should be using the Apollo Client from now on to prepare queries and mutations.
+
+If you are not using the `wp-templates` system, you can take a look a the [search.js](https://github.com/wpengine/atlas-blueprint-portfolio/blob/main/pages/search.js)
+component which uses the next.js `getNextStaticProps` page helpers.
+
+**6. Add the `WordPressTemplate` and `getWordPressProps` when using Template Hierarchy**
+
+For `pages/index.js` and `pages/[...wordpressNode].js`:
+
+```js title="pages/index.js and pages/[...wordpressNode].js"
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:226]:
```suggestion
```jsx title="pages/index.js and pages/[...wordpressNode].js"
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 24]
```diff
@@ -2,6 +2,269 @@
...
+}
+
+export function getStaticProps(ctx) {
+  return getWordPressProps({ ctx });
+}
+```
+
+For `pages/preview.js`:
+
+```js title="pages/preview.js"
```
Comment (blakewilson) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:240]:
```suggestion
```jsx title="pages/preview.js"
```

**Review by blakewilson — COMMENTED** [review]
NIce work @theodesp! Added a few grammar/syntax highlighting fixes.

File: internal/faustjs.org/docs/migrationPath/overview.mdx (hunk 4 above)
Comment (theodesp) [reply internal/faustjs.org/docs/migrationPath/overview.mdx:142]: I think this one is implied.

**Review by blakewilson — APPROVED** [review]
Looks great @theodesp! Killed it 💯

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 25]
```diff
@@ -2,6 +2,279 @@
 slug: /migrationPath/overview
 title: Migration from Legacy Faust
 ---
-Migrate from old version of Faust.js to new Faust
 
-### This document is in progress. Check back soon for content!
\ No newline at end of file
+## General Guidelines
+
+Currently, there is no great automated way to migrate from previous versions of Faust that use GQty to the recent version using Apollo. This is primarily due to their striking differences in the way the queries are performed.
```
Comment (matthewguywright) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:8]: I would make sure the language stays very positive. Maybe change things like "no great automated way" to something like "Currently migration from previous versions of Faust that use GQty is a manual process. To make the process easier we have outlined the necessary steps below."

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 26]
```diff
@@ -2,6 +2,279 @@
...
+However, there are some conventional techniques and best practices for React Development that will definitely help you with this process.
+
+### Reusing Presentational Components
+If you paid attention while developing your Faust website, the Components folder typically contains components that embody a presentational meaning.
```
Comment (matthewguywright) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:12]: I wouldn't say things like "If you paid attention", that seems a little rude or condescending.

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 27]
```diff
@@ -2,6 +2,279 @@
...
+      {date && author && <>&nbsp;</>}
+      {author && <span>By {author}</span>}
+    </div>
+  );
+}
+```
+
+There is no reference to GQty or Apollo and this component can be safely re-used across different applications.
+
+On the other hand the following component is harder to migrate as it depends on the GQty client:
```
Comment (matthewguywright) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:39]:
```suggestion
On the other hand the following component is more difficult to migrate as it depends on the GQty client:
```

File: internal/faustjs.org/docs/migrationPath/overview.mdx [hunk 28]
```diff
@@ -2,6 +2,279 @@
...
+    <nav>...</nav>
+  );
+}
+```
+
+If you developed your components using this approach, you will be able to just transfer them without modifications to the new framework.
+
+### Writing GraphQL Queries
+The biggest hurdle when migrating from GQty to Apollo is that now you will need to write queries manually.
+However this is not as bad as you think. You only need to identify the data that your pages and components need and try to request the appropriate fields.
```
Comment (matthewguywright) [inline internal/faustjs.org/docs/migrationPath/overview.mdx:68]:
```suggestion
You just need to identify the data that your pages and components need and try to request the appropriate fields.
```

**Review by matthewguywright — APPROVED** [review]

**Review by blakewilson — APPROVED** [review]

**Pull request merged**